Refactor serialization IO infrastructure and support HDFS/HTTP

DellCurry opened the pull request:
This is an implementation for #132024.

## Background
Currently, serialization.py does not support IO abstraction. As a result, torch.save and torch.load only support local disk file.
But in most cases, we hope to save the models remotely, e.g. cloud storage or remote server. There are 2 ways to use remote storage at this time:
### download to local
Easy to understand and handle. Every time we hope to save the model, call torch.save, and upload it manually. Similar thing happens when loading a model file from cloud, we have to download the model file to local disk and then call torch.load.
### fuse
FUSE is a userspace filesystem framework. It makes it possible for developer to build a filesystem in user space instead of kernel. [sshfs](https://github.com/libfuse/sshfs) is an example. network filesystem client to connect to SSH servers.

```
                                                                                                +-----------+
USER                                                            +------------------------------>|HDFS cluser|
                                                                |                               +-----------+
        +-------+                                         +-----------+
        |pytorch|                                         |fuse daemon|                         +-----------+
        +-------+                                         +-----------+------------------------>|    S3     |
            |                                                   |                               +-----------+
            | torch.save()                                      | handle fuse write
            |                                                   |
-------------------------------------------------------------------------------------
KERNEL      |                                                   |
            | sys.write()                                       |
	    |                                                   |
	    |                     +--------+                    |
	    +-------------------->|fuse dev|--------------------+
				  +--------+
```

## New IO abstraction
I am going to introduce an IO abstraction for pytorch serialization, also implement 3 new IO drivers:
* file: Compatible with current syntax. Users could save/load models locally. 
* hdfs: Support save/load to HDFS.
* http: Support load from HTTP server (HTTPS on its way).

User can also implement other storage protocols easily,by inheriting base class and implement its own IO methods.

By doing this, we have some advantages:
* Rather than upload/download to local disk file manually, this new framework makes it diskless and convenient.
* Compared to fuse-based filesystem, pytorch read/write data directly. The shorter path reduce cpu and memory of fuse. Besides, we can also avoid pytorch IO stuck in D state (uninterruptable in Linux Kernel) when the network is down.

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

DellCurry committed:
support hdfs storage protocol for torch.save and torch.load

Signed-off-by: zhenwei pi <pizhenwei@bytedance.com>
Signed-off-by: wangyu <wangyu.steph@bytedance.com>

DellCurry committed:
support http protocol for torch.load

Signed-off-by: zhenwei pi <pizhenwei@bytedance.com>
Signed-off-by: wangyu <wangyu.steph@bytedance.com>

DellCurry commented:
Here is an example that fuse-based filesystem got stuck during training. we just encountered this case on Friday.
![20240722-100222](https://github.com/user-attachments/assets/94908335-9ea9-4d59-89b0-259fa91725e5)

mikaylagawarecki reviewed (commented):
Hi @DellCurry, thanks for this PR. For future reference, you could open an issue containing the feature request (mostly like what you had written in the PR description). This would enable us to deliberate on whether we would like to accept this feature before you implement the PR.

Separately some quick questions here:
- Could you elaborate on why `fsspec` is not useful in this case?
- Given that `torch.save` accepts any buffer object that implements `write` and `flush`, while `torch.load` accepts any buffer object that implements `read`, `readline`, `seek` and `tell`, why is it not possible to pass a custom Buffer object that implements the necessary methods (like your `HDFSIO`/`HTTPIO` objects) + references the http/hdfs address to `torch.{save/load}`? Is the problem that an extension point is needed for the concept of "opening" and "closing"?

DellCurry commented:
hi @mikaylagawarecki, thanks for review! Sorry for late reply. I already open an issue #132024 and reference this pr to it.

Here is the explanation for your questions:

Let's start from the second question:

`opening` and `closing` is not the key point. An buffer-like object(implements basic read/write method for remote file) is truly possible. However, if we just implement an HDFSIO, the syntax may be:
```
>>> with hdfs = HDFSIO(host, port, filename):
... torch.load(hdfs, weights_only = True)
```
Obviously, user need one more step to 'open' the HDFSIO.

Besides, since torch.load accepts a file-like object as parameter:
```
FILE_LIKE: TypeAlias = Union[str, os.PathLike, BinaryIO, IO[bytes]]
```
We also need to modify the definition of torch.load, or extend another type of FILE-LIKE.

For the implementation of this PR, user could just pass a url-like string:
```
>>> torch.load('http://host:port/filename', weights_only=True)
```
What we do is just encapsulating the IO module. Beyond just implementing HDFSIO, we also unify the file-open entry. All the file types, including local file, buffer and newly-added protocol url, could all be opened easily by `open_file_like` inside torch.load. It would bring us convenient syntax, and also clear data structure.

Now we go back to the first question about `fsspec`. I am not familiar with this library before. I spent some time reading its code and usage, but still not sure I fully understand it. In my view, it is a library to unify python filesystem. In other words, it is somehow a way to implement HTTPIO (Seems it does not support HDFS). So the answer is similar, we can use `fsspec` to implement some basic IO class (HTTPIO could surely be simplified by it, I would rewrite this part later), but we still need to abstract the IO module.

DellCurry commented:
> Now we go back to the first question about `fsspec`. I am not familiar with this library before. I spent some time reading its code and usage, but still not sure I fully understand it. In my view, it is a library to unify python filesystem. In other words, it is somehow a way to implement HTTPIO (Seems it does not support HDFS). So the answer is similar, we can use `fsspec` to implement some basic IO class (HTTPIO could surely be simplified by it, I would rewrite this part later), but we still need to abstract the IO module.

hi @mikaylagawarecki ! I have rewritten HTTPIO with `fsspec`.

DellCurry commented:
hi @mikaylagawarecki! More ideas on this PR ?

DellCurry committed:
Refactor serialization IO infrastructure

Currently, serialization.py does not support IO abstraction.
As a result, torch.save and torch.load only support local disk file.
These days, most models are saved on cloud storage such as Amazon S3,
HDFS, etc. We must save model to local disk first, then upload it manually.
Similiar thing happens when loading a model file on cloud, we have to
download file to local disk and then call torch.load.

Therefore, we abstract io infrastructure out of serialization.py.
Users can save/load the model directly to any cloud.

Signed-off-by: zhenwei pi <pizhenwei@bytedance.com>
Signed-off-by: wangyu <wangyu.steph@bytedance.com>
